libs/wampcc/wamp_session.cc

```diff
@@ -2707,6 +2723,28 @@ bool wamp_session::has_authid() const
   return m_authid.first;
 }
 
+std::string wamp_session::authrole() const
+{
+  std::lock_guard<std::mutex> guard(m_authrole_lock);
+  return m_authrole;
+}
+
+void wamp_session::authorize(const std::string& uri, auth_provider::action action)
+{
+  if(m_auth_proivder.authorize) {
+    std::lock_guard<std::mutex> guard(m_authrole_lock);
```

Comment:
lock scope here is held during callback to m_auth_provider .... lead to deadlock if subsequence call to session.authrole is made.    I think it should be safe to just take copy of m_authrole .